compiler-plugin/src/main/java/io/ballerina/lib/cdc/compiler/codeaction/AbstractCdcCodeTemplate.java

```diff
@@ -0,0 +1,124 @@
+import io.ballerina.projects.plugins.codeaction.CodeActionContext;
+import io.ballerina.projects.plugins.codeaction.CodeActionExecutionContext;
+import io.ballerina.projects.plugins.codeaction.CodeActionInfo;
+import io.ballerina.projects.plugins.codeaction.DocumentEdit;
+import io.ballerina.tools.diagnostics.Diagnostic;
+import io.ballerina.tools.text.LineRange;
+import io.ballerina.tools.text.TextDocumentChange;
+import io.ballerina.tools.text.TextEdit;
+import io.ballerina.tools.text.TextRange;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import static io.ballerina.lib.cdc.compiler.Constants.IS_POSTGRES_LISTENER;
+import static io.ballerina.lib.cdc.compiler.Constants.NODE_LOCATION;
+import static io.ballerina.lib.cdc.compiler.Utils.findNode;
+
+/**
+ * Abstract class for CDC code templates to share common functionality.
+ */
+public abstract class AbstractCdcCodeTemplate implements CodeAction {
+
+    @Override
+    public List<String> supportedDiagnosticCodes() {
+        return List.of(
+                DiagnosticCodes.EMPTY_SERVICE.getCode(),
+                DiagnosticCodes.EMPTY_SERVICE_POSTGRESQL.getCode()
+        );
+    }
+
+    @Override
+    public Optional<CodeActionInfo> codeActionInfo(CodeActionContext codeActionContext) {
+        Diagnostic diagnostic = codeActionContext.diagnostic();
+        if (diagnostic.location() == null) {
+            return Optional.empty();
+        }
+        CodeActionArgument locationArg = CodeActionArgument.from(NODE_LOCATION, diagnostic.location().lineRange());
+        CodeActionArgument isPostgres = CodeActionArgument.from(IS_POSTGRES_LISTENER,
+                diagnostic.diagnosticInfo().code().equals(DiagnosticCodes.EMPTY_SERVICE_POSTGRESQL.getCode()));
+        return Optional.of(CodeActionInfo.from(getCodeActionDescription(), List.of(locationArg, isPostgres)));
+    }
+
+    protected abstract String getCodeActionDescription();
+
+    @Override
+    public List<DocumentEdit> execute(CodeActionExecutionContext codeActionExecutionContext) {
+        LineRange lineRange = extractLineRange(codeActionExecutionContext);
+        boolean isPostgresListener = extractIsPostgresListener(codeActionExecutionContext);
+
+        if (lineRange == null) {
+            return Collections.emptyList();
+        }
+
+        SyntaxTree syntaxTree = codeActionExecutionContext.currentDocument().syntaxTree();
+        NonTerminalNode node = findNode(syntaxTree, lineRange);
+        if (!(node instanceof ServiceDeclarationNode serviceDeclarationNode)) {
+            return Collections.emptyList();
+        }
+
+        List<TextEdit> textEdits = generateTextEdits(serviceDeclarationNode, isPostgresListener);
+        TextDocumentChange change = TextDocumentChange.from(textEdits.toArray(new TextEdit[0]));
+        return Collections.singletonList(new DocumentEdit(codeActionExecutionContext.fileUri(),
+                SyntaxTree.from(syntaxTree, change)));
```

Comment:
I assume here you want to create an unmodifiable list with only one element. Then we can do this.
```suggestion
        return List.of(new DocumentEdit(codeActionExecutionContext.fileUri(),
                SyntaxTree.from(syntaxTree, change)));
```

Reply:
Fixed in [c936ccb](https://github.com/ballerina-platform/module-ballerinax-cdc/pull/3/commits/c936ccbc4ffbe912c9b9ea28a7cf14acc07f0308)